src/testsuite/vec_f32_dummy.c

```diff
@@ -634,6 +635,175 @@ test_pred_f32_zero (vf32_t value)
 	return (vec_iszerof32 (value));
 }
 
+
+/*
+ * The following are both compile tests for Gather/Scatter operations
+ * and performance kernels for performance tests.
+ */
+
+
+float matrix_f32 [MN][MN] __attribute__ ((aligned (128)));
+
+void
+test_f32_Imatrix_init (float * array)
+{
+  long i, j, k;
+  long rows, columns;
+
+  rows = columns = MN;
+
+#ifdef __DEBUG_PRINT__
+  printf ("init_indentity array[%d,%d]\n",
+	  rows, columns);
+#endif
+
+  for ( i=0; i<rows; i++ )
+  {
+    for ( j=0; j<columns; j++ )
+      {
+	k = (i * columns) + j;
+	if (i == j)
+	  {
+	    array [k] = 1.0;
+#ifdef __DEBUG_PRINT__
+	    printf ("init_indentity array[%d,%d] is %f\n",
+			i, j, array [k]);
+#endif
+	  }
+	else
+	  {
+	    array [k] = 0.0;
+	  }
+      }
+  }
+}
+
+void
+#if !defined(__clang__)
+__attribute__ ((optimize ("unroll-loops")))
+#endif
+test_f32_matrix_transpose (float * tm, float * m)
+{
+  long i, j, k, l;
+  long rows, columns;
+
+  rows = columns = MN;
+
+  for ( i=0; i<rows; i++ )
+  {
+    for ( j=0; j<columns; j++ )
+      {
+	k = (i * columns) + j;
+	l = (j * columns) + i;
+	tm[l] = m[k];
+      }
+  }
+}
+
+void
+//__attribute__ ((optimize ("unroll-loops")))
```

Comment:
Why some functions do not have this attribute? Is unroll-loops not giving any performance enhancement?

Reply:
Clang can't handle this attribute and and requires a conditional to avoid compile error.
i.e. 
#if !defined(__clang__)
__attribute__ ((optimize ("unroll-loops")))
#endif

The basic c loop matrix transpose does benefit from "unroll-loops" with GCC, while the "gather" transpose versions not so much. This is likely due to the the gather implementation using in-line asm which limits code motion.

This comment serves as a reminder to look at this again, later.

Reply:
May I suggest a different solution, then?
I think it could reduce code duplication and make it easier when clang starts supporting this attribute.

It could go like this:

```
#if !defined clang
# define UNROLL_LOOPS attribute ((optimize ("unroll-loops")))
#else
# define UNROLL_LOOPS
#endif

void UNROLL_LOOPS
foo(...)
{
...
}

int UNROLL_LOOPS
bar(...)
{
...
}
```